src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -333,40 +353,48 @@ private void InnerStartOperationReceiveFrom()
 
         internal unsafe SocketError DoOperationReceiveFrom(SafeCloseSocket handle, out SocketFlags flags)
         {
-            PrepareIOCPOperation();
-
             flags = _socketFlags;
 
-            int bytesTransferred;
-            SocketError socketError;
-            if (_buffer != null)
+            SocketError socketError = SocketError.Success;
+            NativeOverlapped* overlapped = AllocateNativeOverlapped();
+            try
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    ref _wsaBuffer,
-                    1,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                int bytesTransferred;
+
+                if (_buffer != null)
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        ref _wsaBuffer,
+                        1,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+                else
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        _wsaBufferArray,
+                        _bufferListInternal.Count,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+
+                socketError = ProcessIOCPResult(socketError == SocketError.Success, bytesTransferred);
+                return socketError;
             }
-            else
+            finally
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    _wsaBufferArray,
-                    _bufferListInternal.Count,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                FreeNativeOverlappedIfNotPending(overlapped, socketError);
```

Comment:
> To match the SocketPal pattern, you could pass the NativeOverlapped* to ProcessIOCPResult and free it on sync completion

That's what I had initially (and is what I meant by passing the overlapped around), but in ProcessIOCPResult I then had to insert the Free call at a couple of different places, and I was also concerned there may be some other points where an exception could emerge and cause it not to be freed.  Previously that would have at least been handled by the SocketAsyncEventArgs.Dispose, because it would have freed the stored overlapped then, but with this change, it would simply be leaked.  So I opted for the finally approach.

If it ends up mattering, we can revisit.